doc/format-directory.md

```diff
@@ -0,0 +1,178 @@
+% LSM specification: directory and file layouts
+% Duncan Coutts
+
+# Scope
+
+This document is intend to cover the overall LSM directory format, including:
+
+ * persistent snapshots
+ * live handles
+ * sufficient checksum/consistency info to detect directory corruption
+   (not just individual file content corruption)
+ * high-level configuration options persisted to files (if necessary)
+
+This document is part of a group of documents that cover the formats:
+ * `format-directory.md`: covering the overall LSM directory format
+ * `format-run.md`: covering the LSM file-run format
+ * `format-page.md`: covering the LSM file-run _page_ format
+
+
+# Top level layout
+
+The filesystem representation for an LSM session lives under a single directory
+structure. We will describe all files relative to this `${session}` root.
+
+The top level directory contains
+
+ * `${session}/lock` file
+ * `${session}/active/` directory
+ * `${session}/snapshots/` directory
+
+The top level lock file is used with an OS file lock API to (cooperatively)
+ensure exclusive use of a session. It is otherwise empty.
+
+# Active directory layout
+
+The ephemeral directory contains all of the files for all the LSM runs that are
+in use by any open LSM handle in the session.
+
+ * `${session}/active/${n}.keyops`
+ * `${session}/active/${n}.filter`
+ * `${session}/active/${n}.index`
+ * `${session}/active/${n}.checksums`
+
+The LSM runs are numbered using a session-wide counter that is initialised from
+0 each time a session is opened.
+
+Note that there are no metadata files in the active directory (as there are for
+snapshots). This is because the metadata is held in memory for open LSM handles.
+
+# Snapshots directory layout
+
+Each snapshot has a name in a session-wide namespace, and gets a corresponding
+subdirectory:
+
+ * `${session}/snapshots/${name}/` directory
+
+Thus the limitations on portable file/directory names applies to snapshot names.
+
+Each snapshot contains a metadata file for the snapshot overall and a checksum:
+
+ * `${session}/snapshots/${name}/snapshot`
+ * `${session}/snapshots/${name}/snapshot.checksum`
+
+plus the five files for each LSM run in the snapshot:
+
+ * `${session}/snapshots/${name}/${n}.keyops`
+ * `${session}/snapshots/${name}/${n}.blobs`
+ * `${session}/snapshots/${name}/${n}.filter`
+ * `${session}/snapshots/${name}/${n}.index`
+ * `${session}/snapshots/${name}/${n}.checksum`
+
+In this case the LSM run files are numbered from 0 within each snapshot.
+
+## Snapshot metadata
+
+The snapshot metadata file contains the following information:
+ * A format version/tag identifier
+ * The stored parameters for the LSM table (specified when the table was
+   originally created), including
+   - LSM tuning parameters
+   - Page size for all the key/operations files
+   - Index type and parameters
+ * Key/value type information for dynamic-type sanity checking
```

Comment:
What type of info would fall into this category? I'm not sure what you mean by dynamic-type sanity checking.

Reply:
types of the keys & values (& blobs)